New telepresence replace command.

thallgren opened the pull request:
The new `telepresence replace` command simplifies and clarifies container replacement.

Previously, the `--replace` flag within the `telepresence intercept` command was used to replace containers. However, this approach introduced inconsistencies and limitations:

* **Confusion:** Using a flag to modify the core function of a command designed for traffic interception led to ambiguity.
* **Inaccurate Behavior:** Replacement was not possible when no incoming traffic was intercepted, as the command's design focused on traffic routing.

To address these issues, the `--replace` flag within `telepresence intercept` has been deprecated. The new `telepresence replace` command provides a dedicated and consistent method for replacing containers, enhancing clarity and reliability.

Key differences between `replace` and `intercept`:

1. **Scope:** The `replace` command targets and affects an entire container, impacting all its traffic, while an `intercept` targets specific services and/or service/container ports.
2. **Port Declarations:** Remote ports specified using the `--port` flag are container ports.
3. **No Default Port:** A `replace` can occur without intercepting any ports.
4. **Container State:** During a `replace`, the original container is no longer active within the cluster.

The deprecated `--replace` flag still works, but is hidden from the `telepresence intercept` command help, and will print a deprecation warning when used.

thallgren commented:
Still a draft. Integration and docs are in the works.

thallgren committed:
Base connectedSuite on TrafficManager instead of NamespacePair.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
New telepresence replace command.

The new `telepresence replace` command simplifies and clarifies
container replacement.

Previously, the `--replace` flag within the `telepresence intercept`
command was used to replace containers. However, this approach
introduced inconsistencies and limitations:

* **Confusion:** Using a flag to modify the core function of a command
  designed for traffic interception led to ambiguity.
* **Inaccurate Behavior:** Replacement was not possible when no incoming
  traffic was intercepted, as the command's design focused on traffic
  routing.

To address these issues, the `--replace` flag within `telepresence
intercept` has been deprecated. The new `telepresence replace` command
provides a dedicated and consistent method for replacing containers,
enhancing clarity and reliability.

Key differences between `replace` and `intercept`:

1. **Scope:** The `replace` command targets and affects an entire
   container, impacting all its traffic, while an `intercept` targets
   specific services and/or service/container ports.
2. **Port Declarations:** Remote ports specified using the `--port` flag
   are container ports.
3. **No Default Port:** A `replace` can occur without intercepting any
   ports.
4. **Container State:** During a `replace`, the original container is no
   longer active within the cluster.

The deprecated `--replace` flag still works, but is hidden from the
`telepresence intercept` command help, and will print a deprecation
warning when used.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Using the --namespace option with telepresence causes a deadlock.

Using `telepresence list --namespace <ns> with a namespace different
from the one that telepresence was connected to, would cause a deadlock,
and then produce an empty list.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Fix problem with exclude-suffix being hidden by DNS search path.

In some situations, a name ending with an exclude-suffix like "xyz.com"
would be expanded by a search path into "xyz.com.<connected namespace>"
and therefore not be excluded. Instead, the name was sent to the cluster
to be resolved, causing an unnecessary load on its DNS server.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Ensure that the correct pod-log is captured after a helm rollback.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Integration tests for the replace command.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Use Serve function of grpc.Server in favor of dhttp.ServerConfig.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
General overhaul of client daemon quit logic.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Remove some excessive debug logging.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Clean up traffic-manager code dealing with session termination.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Fix some issues with pod-scaling integration test.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Let kubeconfig_extension_test.go use proper client config.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Ensure that generic.goyaml can use fully qualified image name.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
We must use podIP to identify agents. StatefulSet will reuse pod names.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Reuse const for "traffic-manager".

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
DNS optimizations on the traffic-manager side.

Let traffic-manager short-circuit DNS requests to itself.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Fix mount sync problem when intercept ended.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Introduce explicit SessionID type.

Also removes the `sessions` map from the traffic-manager's `state`
because it has been made redundant.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Use a GRPC interceptor for logging.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Better error handling in FindOwnerWorkload.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Drop use of the telepresence-agents configmap.

This commit removes the complex pattern used when engaging with a
workload to get a traffic-agent injected into its pods.

1. Generate an agent configuration.
2. Store the configuration in the `telepresence-agents` configmap.
3. Wait for the agent to arrive.

A watcher on the `telepresence-agents` configmap would now get an event
and discover the new or modified configuration. It took the following
steps:
1. Determine if a rollout is needed.
2. Perform a rollout to trigger the mutating webhook.

1. Check if the pod has a configuration entry in the configmap.
2. Patch pod if necessary.

1. Generate an agent configuration.
2. Evict pods with config annotation mismatch (triggers the webhook).

No action. The configmap no longer exists

1. Check if the pod has a configuration entry in the configmap.
2. Patch pod if necessary and include the config as an annotation.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Introduce k8sapi.Kind and use it consistently instead of string.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Use podUid to identify pods.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Merge the traffic-manager interceptStates and intercepts maps.

The split between the two was for historical reasons. The `intercepts`
map was only capable of storing proto-bufs. Since that's no longer the
case, those two maps can be merged.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Minor fixes to integration tests.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Update module dependencies.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Fix panic when allowing conflicting subnets on macOS.

This commit fixes a regression introduced in 2.21.0, causing the macOS
client to panic when the `--allow-conflicting-subnets` connect flag was
used, or when `routing.allowConflictingSubnets` was added to the client
configuration.

Closes #3784

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
release workflow fix

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Using the --proxy-via flag would sometimes cause connection timeouts.

Typically, a `telepresence connect --proxy-via <subnet>=<workflow>`
would fail with a "deadline exceeded" message when several workloads
were present in the namespace, the one targeted by the proxy-via didn't
yet have an agent installed, and other workloads had an agent. This was
due to a race condition in the logic for the agent-based port-forwards
in the root daemon. The conditions causing this race are now eliminated.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Update actions/upload-artifact to v4.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Ensure that make generate regenerates the docs files.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Prepare v2.21.3-test.4

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Ensure that annotation enabled traffic-agents are uninstalled.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Merge pull request #3785 from telepresenceio/thallgren/v2.21.3

Fix panic when allowing conflicting subnets on macOS.

thallgren committed:
Add build flag to link the go-fuseftp implementation.

This build-flag is not enabled by default, but can be used when there
is a desire to build a telepresence client that links the fuseftp
implementation instead of using a remote process via gRPC.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Prefer route with gateway when adding static route.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Merge pull request #3786 from telepresenceio/thallgren/proxy-via-multi-wf

Using the --proxy-via flag would sometimes cause connection timeouts.

thallgren committed:
Never chose route with different IP family.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Prepare v2.21.3-test.5

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Merge pull request #3790 from telepresenceio/thallgren/macos-allow-conflicts

Prefer route with gateway when adding static route.

thallgren committed:
Prepare v2.21.3-rc.0

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Get rid of the namespace lock mechanism. No longer needed.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Rename "delete" pod methods to "evict" to clarify.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Prepare v2.21.3

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Merge branch 'release/v2.21' into release/v2

thallgren committed:
Safeguard concurrent access to agentpf.client with RWMutex.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

P0lip commented on pkg/workload/watcher.go:
```diff
@@ -106,19 +94,13 @@ func NewWatcher(ctx context.Context, ns string, enabledWorkloadKinds []Kind) (Wa
-
-			case "Rollout":
-				if slices.Contains(enabledKinds, RolloutKind) {
-					return true
-				}
+			kind := k8sapi.Kind(ref.Kind)
+			switch kind {
+			case k8sapi.DeploymentKind, k8sapi.RolloutKind:
+				return enabledKinds.Contains(kind)
```
this introduces a tiny change in the behavior that might be unwanted.
Previously, if the kind wasn't listed as enabled, we'd continue the iteration as possibly there would be another kind that's enabled.
In other words, the old behavior would be


```go
  if enabledKinds.Contains(kind) {
    return true
  }
```

thallgren replied:
There's no point in continuing the iteration. An object can only have one controller, and once the condition on line 111/99 is true, that controller has been found.

thallgren committed:
Add better debug logging around the agent pod watcher mechanism.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Don't let the agent pod watcher wait for port-forward to be established

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Introduce a proper tunnel.Tag to make some sense out of trace logging.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Ensure that tunnel write-delays cause connection back-pressure.

When reading from a connection and then writing the data on a tunnel,
it is essential that any delay when writing also causes the next read
from the connection to be delayed. Delaying the connection read causes
a back-pressure on the TCP stack, and will make the sender aware that
things need to slow down, which is exactly what we want.

Closes #3766

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Use the default TCP window and 1024 maxInFlight.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Print stack-traces on debug-level if session cancellation hangs.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Slightly more efficient echo-server.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Really massage concurrency in Test_LargeRequest.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Ensure removal of PV and PVC in integration tests.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Close root daemon connection to traffic-manager early on disconnect.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Let conn<->stream dispatcher close reader and writer of conn separately

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
List output includes workload kind.

The output of the `telepresence list` command will now include the
workload kind (deployment, replicaset, statefulset, or rollout) in all
entries.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Documentation changes for the new replace command.

Introducing `telepresence replace` as an alternative to the `intercept`
resulted in a lot of documentation changes. For many (most) use-cases,
a `replace` is more natural than an `intercept` unless there's
an ability to do personal intercepts, and the OSS edition doesn't
provide that.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Let make docs-files target generate docs/variables.yaml from changelog.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Ensure that agent watchers don't receive inactivated agents.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Remove all panic recovery.

Most of the panic recovery routines were added to ensure logging the
stack-trace of the panic. Containers in pods seem to run with
`GOTRACEBACK=""` which apparently overrides the default "single"
setting. To mitigate this, the traffic-manager, traffic-agent, and
agent-init containers now use the `debug.SetTraceback("single")`

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Fixup and enable Test_LargeFiles for CI.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

P0lip commented on docs/reference/intercepts/cli.md:
I'm not sure how the hosted documentation URLs are built, but if they are inferred from the directory structure, we could rename the `intercepts` directory here

P0lip commented on cmd/traffic/cmd/manager/mutator/service.go:
we still have panic recovery in we still have `serveMutatingFunc` - didn't you intend to remove it in https://github.com/telepresenceio/telepresence/commit/59fb3f70b84b9bb935308205a85cf3d3a0eb2f4c#diff-e6df1931e03199997ae8c5c0ec6325293de7862dd39912c2995d2099cfa06032?

P0lip commented on pkg/client/cli/cmd/genyaml.go:
```diff
@@ -421,6 +419,55 @@ func (g *genInitContainerInfo) run(cmd *cobra.Command, kubeFlags map[string]stri
+	}
+	flags := cmd.Flags()
+	flags.StringVarP(&info.workloadName, "workload", "w", "",
+		"Name of the workload. If given, the configmap entry will be retrieved telepresence-agents configmap, mutually exclusive to --config")
```
```suggestion
		"Name of the workload. If given, the configmap entry will be retrieved from the telepresence-agents configmap, mutually exclusive to --config")
```

P0lip commented on integration_test/workloads_test.go:
```diff
@@ -24,6 +24,7 @@ func (s *connectedSuite) successfulIntercept(tp, wl, port string) {
 		2*time.Second, // polling interval
 	)
 
+	itest.TelepresenceOk(ctx, "loglevel", "trace")
```
It'd be good to restore to the previous state
```suggestion
	itest.TelepresenceOk(ctx, "loglevel", "trace")
	defer itest.TelepresenceOk(ctx, "loglevel", "debug")
```

P0lip reviewed (approved):
Nice, I like the removal of tp agents configmap

thallgren committed:
Rename docs/references/intercepts to engagements.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Changes in response to code review.

- Remove panic recovery in mutator/service.go
- Remove extraneous loglevel setting in succesfulIntercept itest
- Remove unused `--workload` flag from some genyaml commands.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Let test-report output append to already existing test.log file.

Before this commit, the test-report command would clobber the output
file. That's bad when a test is repeated due to failure.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Improve deduction of workload state.

Signed-off-by: Thomas Hallgren <thomas@tada.se>
